src/main/java/net/neoforged/neoforge/transfer/transaction/TransactionManagerImpl.java

```diff
@@ -0,0 +1,227 @@
+package net.neoforged.neoforge.transfer.transaction;
+
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ArrayList;
+
+@ApiStatus.Internal
+public class TransactionManagerImpl {
+    public static final ThreadLocal<TransactionManagerImpl> MANAGERS = ThreadLocal.withInitial(TransactionManagerImpl::new);
+
+    private final Thread thread = Thread.currentThread();
+    private final ArrayList<TransactionImpl> stack = new ArrayList<>();
+    private final ArrayList<Transaction.OuterCloseCallback> outerCloseCallbacks = new ArrayList<>();
+    private int currentDepth = -1;
+
+    public boolean isOpen() {
+        return currentDepth > -1;
+    }
+
+    public Transaction openOuter() {
+        if (isOpen()) {
+            throw new IllegalStateException("An outer transaction is already active on this thread.");
+        }
+
+        return open();
+    }
+
+    @Nullable
+    public TransactionContext getCurrentUnsafe() {
+        if (currentDepth == -1) {
+            return null;
+        } else if (stack.get(currentDepth).lifecycle == Transaction.Lifecycle.OPEN) {
```

Comment:
If we are actually going to have this method, we should at least flatten things a bit so that it doesn't have to do two `get` calls rather than doing them just to allow for chaining else if statements